0000-ledger-plugin-interface-2/0000-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,510 @@
+
+#### Event: `error`
+<code>ledgerPlugin.on('error', ( **err**:Error ) ⇒ )</code>
+
+General event for fatal exceptions. Emitted when the plugin experienced an unexpected unrecoverable condition. Once triggered, this instance of the plugin MUST NOT be used anymore.
+
+### Ledger Transfers
+
+#### LedgerPlugin#sendTransfer
+<code>ledgerPlugin.sendTransfer( **transfer**:[Transfer](#class-transfer) ) ⇒ Promise.&lt;[FulfillmentInfo](#class-fulfillmentinfo)></code>
+
+Initiates an account-local transfer. A transfer MUST contain an `amount` of zero or more and MAY have attached `ilp` data. See the description of the [Transfer](#class-transfer) class below.
+
+All plugins MUST support amounts in a range from zero to some maximum. Plugins MAY implement zero-amount transfers differently than other transfers.
+
+###### Parameters
+| Name | Type | Description |
+|:--|:--|:--|
+| transfer | <code>[Transfer](#class-transfer)</code> | Properties of the transfer to be created |
+
+When sending transfers, all fields are required and MUST be provided, even if they contain an empty value, such as an empty object in `custom`.
+
+###### Returns
+**`Promise.<FulfillmentInfo>`** A promise which resolves when the transfer has been fulfilled/executed.
+
+Rejects with `InvalidFieldsError` if required fields are missing from the transfer or malformed. Rejects with `NotAcceptedError` if the transfer is rejected by the ledger due to insufficient balance or any other reason. Rejects with `NotConnectedError` if the plugin is not connected to the ledger.
+
+Rejects with [`InterledgerRejectionError`](#class-interledgerrejectionerror) if the other side rejects the transfer and attaches ILP rejection data. Rejects with [`GenericRejectionError`](#class-genericrejectionerror) if the other side rejects the transfer, but does not attach valid ILP rejection data. One example for when a GenericRejectionError may be used is if a transfer times out.
+
+This method MAY reject with any arbitrary JavaScript error.
+
+###### Example
+```js
+p.sendTransfer({
+  amount: '10',
+  executionCondition: Buffer.from('47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU', 'base64'),
+  expiresAt: '2016-05-18T12:00:00.000Z',
+  ilp: Buffer.alloc(0),
+  custom: {}
+})
+```
+
+For a detailed description of these properties, please see [`Class: Transfer`](#class-transfer).
+
+#### LedgerPlugin#registerTransferHandler
+<code>ledgerPlugin.registerTransferHandler( **transferHandler**: ( transfer: [Transfer](#class-transfer) ) ⇒ Promise&lt;[FulfillmentInfo](#class-fulfillmentinfo)> ) ⇒ null</code>
+
+Set the callback which is used to handle incoming prepared transfers. The callback should expect one parameter (the [Transfer](#class-transfer)) and return a promise for the resulting [FulfillmentInfo](#class-fulfillmentinfo). If the transfer is rejected or an error occurs, the callback should reject the transfer. In general, the callback should behave as [`sendTransfer`](#ledgerpluginsendtransfer) does.
```

Comment:
@justmoon While working on the PSKv2 implementation using the LPI2 I realized we also want to have a way to ignore an incoming transfer. This is necessary if you have two receivers listening on one account and want to make sure they don't interfere with one another.

Reply:
actually @justmoon and I were briefly discussing this yesterday. It might be a best practice to have each listener get its own sub-account. This means your ILP address would change every time you reconnect (which is OK because they're ephemeral anyways), but all your listeners would still have a shared balance. The ILP-DCP protocol would just have to take into account which listener is asking, and then send out to that same listener when notifications come in.